code/game/machinery/computer/apc_control.dm

```diff
@@ -0,0 +1,208 @@
+	var/datum/browser/popup = new(user, "apc_control", name, 600, 400)
+	popup.set_content(dat)
+	popup.set_title_image(user.browse_rsc_icon(icon, icon_state))
+	popup.open()
+
+/obj/machinery/computer/apc_control/Topic(href, href_list)
+	if(..())
+		return
+	var/image/I = image(src) //For feedback message flavor
+	if(href_list["authenticate"])
+		var/obj/item/weapon/card/id/ID = usr.get_active_held_item()
+		if(!istype(ID))
+			ID = usr.get_idcard()
+		if(ID && istype(ID))
+			if(check_access(ID))
+				authenticated = TRUE
+				auth_id = "[ID.registered_name] ([ID.assignment])"
+				log_activity("logged in")
+		if(!authenticated) //Check for emags
+			var/obj/item/weapon/card/emag/E = usr.get_active_held_item()
+			if(E && istype(E))
+				usr << "<span class='warning'>You bypass [src]'s access requirements using your emag.</span>"
+				authenticated = TRUE
+				log_activity("logged in") //Auth ID doesn't change, hinting that it was illicit
+	if(href_list["log_out"])
+		log_activity("logged out")
+		authenticated = FALSE
+		auth_id = "\[NULL\]"
+	if(href_list["restore_logging"])
+		to_chat(usr, "<span class='robot notice'>\icon[I] Logging functionality restored from backup data.</span>")
+		emagged = FALSE
+		LAZYADD(logs, "<b>-=- Logging restored to full functionality at this point -=-</b>")
+	if(href_list["access_apc"])
+		playsound(src, "terminal_type", 50, 0)
+		var/obj/machinery/power/apc/APC = locate(href_list["access_apc"]) in apcs_list
+		if(!APC || APC.aidisabled || APC.panel_open)
+			to_chat(usr, "<span class='robot danger'>\icon[I] APC does not return interface request. Remote access may be disabled.</span>")
+			return
+		if(active_apc)
+			to_chat(usr, "<span class='robot danger'>\icon[I] Disconnected from [active_apc].</span>")
+			active_apc.say("Remote access canceled. Interface locked.")
+			playsound(active_apc, 'sound/machines/BoltsDown.ogg', 25, 0)
+			playsound(active_apc, 'sound/machines/terminal_alert.ogg', 50, 0)
+			active_apc.locked = TRUE
+			active_apc.update_icon()
+			active_apc = null
+		to_chat(usr, "<span class='robot notice'>\icon[I] Connected to APC in [get_area(APC)]. Interface request sent.</span>")
+		log_activity("remotely accessed APC in [get_area(APC)]")
+		APC.interact(usr, not_incapacitated_state)
+		playsound(src, 'sound/machines/terminal_prompt_confirm.ogg', 50, 0)
+		message_admins("[key_name_admin(usr)] remotely accessed [APC] from [src] at [get_area(src)].")
+		log_game("[key_name_admin(usr)] remotely accessed [APC] from [src] at [get_area(src)].")
+		if(APC.locked)
+			APC.say("Remote access detected. Interface unlocked.")
+			playsound(APC, 'sound/machines/BoltsUp.ogg', 25, 0)
+			playsound(APC, 'sound/machines/terminal_alert.ogg', 50, 0)
+		APC.locked = FALSE
+		APC.update_icon()
+		active_apc = APC
+	if(href_list["name_filter"])
+		playsound(src, 'sound/machines/terminal_prompt.ogg', 50, 0)
+		var/new_filter = stripped_input(usr, "What name are you looking for?", name) as null|text
```

Comment:
these are sleeps, so you need to check the item isn't qdeleted and all the assumptions are still valid (emag valid, player still alive and not deleted etc)

putting all the checks in canUseTopic and then calling that in all the plcaes would be good